sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -1307,7 +1308,16 @@ class Dataset[T] private[sql](
    */
   @scala.annotation.varargs
   def hint(name: String, parameters: Any*): Dataset[T] = withTypedPlan {
-    UnresolvedHint(name, parameters, logicalPlan)
+    if (ResolveHints.ResolveCoalesceHints.COALESCE_HINT_NAMES.contains(
```

Comment:
Ur, I see.... I feel a bit weird handling here... we might be better to give up some representations (e.g., `hint("repartition", 1, df("a"))`, `hint("repartition", 1, $"a")`, ....) so as to simplify the implementation. cc: @maryannxue @dongjoon-hyun

Reply:
On second thoughts, its ok to drop the support of the two patterns above for simplification in this pr. If we need more rich syntaxes, we can do it in follow-ups. Could you do it and update the PR description: which syntax is supported or not in this pr?

Reply:
> We need .expr? I think this is not intuitive for users... Can you try #25464 (comment)?

As you said, if you do not support this syntax, user may feel strange.  How about throw a kind Exception, like this
```
if (ResolveHints.ResolveCoalesceHints.COALESCE_HINT_NAMES.contains( name.toUpperCase(Locale.ROOT))) {
    throw new IllegalArgumentException(s"For coalesce hint , Please use $name(...) instead")
}
```

Reply:
I feel the change in Dataset.hint is a little intrusive, so I don't touch the current code as much as possible... cc: @HyukjinKwon

Reply:
Thank you for tracking this pr.
Originally, this pr aims to support partition by hint for spark sql command. So if it is more better that revert the `dataset.hint` and just leave it to an another pr ?

Reply:
Oh, I missed here. Yea, can we avoid to handle here?

Reply:
OK, revert it.